libraries/networking/src/udt/Socket.h

```diff
@@ -37,34 +44,49 @@ class Socket : public QObject {
     
     quint16 localPort() const { return _udpSocket.localPort(); }
     
-    qint64 writeUnreliablePacket(const Packet& packet, const HifiSockAddr& sockAddr);
-    
-    qint64 writeDatagram(const char* data, qint64 size, const HifiSockAddr& sockAddr)
-        { return writeDatagram(QByteArray::fromRawData(data, size), sockAddr); }
+    // Simple functions writing to the socket with no processing
+    qint64 writeBasePacket(const BasePacket& packet, const HifiSockAddr& sockAddr);
+    qint64 writePacket(const Packet& packet, const HifiSockAddr& sockAddr);
+    qint64 writePacket(std::unique_ptr<Packet> packet, const HifiSockAddr& sockAddr);
+    qint64 writeDatagram(const char* data, qint64 size, const HifiSockAddr& sockAddr);
     qint64 writeDatagram(const QByteArray& datagram, const HifiSockAddr& sockAddr);
     
-    void bind(const QHostAddress& address, quint16 port = 0) { _udpSocket.bind(address, port); }
+    void bind(const QHostAddress& address, quint16 port = 0) { _udpSocket.bind(address, port); setSystemBufferSizes(); }
     void rebind();
     
     void setPacketFilterOperator(PacketFilterOperator filterOperator) { _packetFilterOperator = filterOperator; }
     void setPacketHandler(PacketHandler handler) { _packetHandler = handler; }
     
-    void setBufferSizes(int numBytes);
-    
     void addUnfilteredHandler(const HifiSockAddr& senderSockAddr, BasePacketHandler handler)
         { _unfilteredHandlers[senderSockAddr] = handler; }
+    
+    void setCongestionControlFactory(std::unique_ptr<CongestionControlVirtualFactory> ccFactory);
+    
+    void connectToSendSignal(const HifiSockAddr& destinationAddr, QObject* receiver, const char* slot);
+    
+    ConnectionStats::Stats sampleStatsForConnection(const HifiSockAddr& destination);
+    std::vector<HifiSockAddr> getSockAddr();
 
 private slots:
     void readPendingDatagrams();
+    void rateControlSync();
     
 private:
+    void setSystemBufferSizes();
+    Connection& findOrCreateConnection(const HifiSockAddr& sockAddr);
+    
     QUdpSocket _udpSocket { this };
     PacketFilterOperator _packetFilterOperator;
     PacketHandler _packetHandler;
     
     std::unordered_map<HifiSockAddr, BasePacketHandler> _unfilteredHandlers;
-
-    std::unordered_map<HifiSockAddr, Packet::SequenceNumber> _packetSequenceNumbers;
+    std::unordered_map<HifiSockAddr, SequenceNumber> _unreliableSequenceNumbers;
+    std::unordered_map<HifiSockAddr, std::unique_ptr<Connection>> _connectionsHash;
```

Comment:
The more I watch c++ talks about performance and optimization the more weary I am of using anything other than `std::vector`. It goes against everything I've ever done, considering I've almost always chosen data structures based on how I will be using them and their time complexity. It would be fun to actually profile this at some point, especially because this is a data structure that could be accessed a lot.